File: x402/crypto.py
"""Casper Ed25519 key handling for x402 payment signatures.

A Casper *Ed25519* account public key is the single byte ``0x01`` followed by
the 32-byte raw Ed25519 public key, hex-encoded (66 hex chars). The PayMesh
``casper-exact`` x402 scheme signs a canonical authorization string with the
sender's Ed25519 private key; the facilitator verifies it against the ``from``
account.
"""

from __future__ import annotations

import secrets
from dataclasses import dataclass

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.exceptions import InvalidSignature

ED25519_PREFIX = "01"  # Casper account-kind tag for Ed25519 keys.


@dataclass
class Account:
    """A Casper Ed25519 account, with its private key for signing."""

    private_key_hex: str
    public_account_hex: str  # "01" + 32-byte raw pubkey hex
    label: str = ""

    @property
    def public_key_hex(self) -> str:
        """Raw 32-byte public key hex (without the 0x01 tag)."""
        return self.public_account_hex[2:]


def generate_account(label: str = "") -> Account:
    """Create a fresh Ed25519 Casper account."""
    priv = Ed25519PrivateKey.generate()
    raw_pub = priv.public_key().public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )
    priv_hex = priv.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),
    ).hex()
    return Account(
        private_key_hex=priv_hex,
        public_account_hex=ED25519_PREFIX + raw_pub.hex(),
        label=label,
    )


def account_from_private_key(private_key_hex: str, label: str = "") -> Account:
    """Reconstruct an :class:`Account` from a raw 64-byte private key hex."""
    priv = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(private_key_hex))
    raw_pub = priv.public_key().public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )
    return Account(
        private_key_hex=private_key_hex,
        public_account_hex=ED25519_PREFIX + raw_pub.hex(),
        label=label,
    )


def canonical_authorization(
    sender: str, recipient: str, value: str, service_id: str, nonce: str
) -> str:
    """The exact string the sender signs."""
    return f"{sender}\n{recipient}\n{value}\n{service_id}\n{nonce}"


def sign_message(message: str, private_key_hex: str) -> str:
    """Sign ``message`` with a raw Ed25519 private key; return hex signature."""
    priv = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(private_key_hex))
    return priv.sign(message.encode("utf-8")).hex()


def verify_signature(
    message: str, signature_hex: str, public_account_hex: str
) -> bool:
    """Verify a signature against a Casper Ed25519 account (``01``-prefixed)."""
    try:
        raw_pub_hex = (
            public_account_hex[2:]
            if public_account_hex.startswith(ED25519_PREFIX)
            else public_account_hex
        )
        pub = Ed25519PublicKey.from_public_bytes(bytes.fromhex(raw_pub_hex))
        pub.verify(bytes.fromhex(signature_hex), message.encode("utf-8"))
        return True
    except (ValueError, InvalidSignature):
        return False


def new_nonce() -> str:
    """A fresh single-use nonce for replay protection."""
    return secrets.token_hex(16)

File: sdk/js/src/crypto.ts
// Casper Ed25519 account handling for x402 payment signatures.
// A Casper Ed25519 account public key = 0x01 tag + 32-byte raw pubkey, hex.
// Uses @noble/ed25519 (pure JS, no native deps, works in browser & Node).

import * as ed from "@noble/ed25519";
import { sha512 } from "@noble/hashes/sha512";

// @noble/ed25519 v2 requires a SHA-512 backend to be wired in.
const _sha512Sync = (...msgs: Uint8Array[]) => sha512(ed.etc.concatBytes(...msgs));
ed.etc.sha512Sync = _sha512Sync as any;
ed.etc.sha512Async = async (...msgs: Uint8Array[]) => _sha512Sync(...msgs);

export const ED25519_PREFIX = "01";

export interface Account {
  privateKeyHex: string;
  publicAccountHex: string; // "01" + 32-byte raw pubkey hex
  label?: string;
}

function toHex(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += b.toString(16).padStart(2, "0");
  return s;
}

function fromHex(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(hex.substr(i * 2, 2), 16);
  return out;
}

export async function generateAccount(label?: string): Promise<Account> {
  const priv = ed.utils.randomPrivateKey();
  return accountFromPrivateKey(toHex(priv), label);
}

export async function accountFromPrivateKey(
  privateKeyHex: string,
  label?: string
): Promise<Account> {
  const pub = await ed.getPublicKey(fromHex(privateKeyHex));
  return {
    privateKeyHex,
    publicAccountHex: ED25519_PREFIX + toHex(pub),
    label,
  };
}

export function canonicalAuthorization(
  sender: string,
  recipient: string,
  value: string,
  serviceId: string,
  nonce: string
): string {
  return `${sender}\n${recipient}\n${value}\n${serviceId}\n${nonce}`;
}

export async function signMessage(
  message: string,
  privateKeyHex: string
): Promise<string> {
  const sig = await ed.signAsync(
    new TextEncoder().encode(message),
    fromHex(privateKeyHex)
  );
  return toHex(sig);
}

export async function verifySignature(
  message: string,
  signatureHex: string,
  publicAccountHex: string
): Promise<boolean> {
  try {
    const rawHex = publicAccountHex.startsWith(ED25519_PREFIX)
      ? publicAccountHex.slice(2)
      : publicAccountHex;
    return await ed.verifyAsync(
      fromHex(signatureHex),
      new TextEncoder().encode(message),
      fromHex(rawHex)
    );
  } catch {
    return false;
  }
}

export function newNonce(): string {
  const bytes = new Uint8Array(16);
  crypto.getRandomValues(bytes);
  return toHex(bytes);
}
